sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -340,11 +370,48 @@ case class TransformWithStateExec(
     )
   }
 
-  override def validateAndMaybeEvolveStateSchema(hadoopConf: Configuration): Unit = {
-    // TODO: transformWithState is special because we don't have the schema of the state directly
-    // within the passed args. We need to gather this after running the init function
-    // within the stateful processor on the driver. This also requires a schema format change
-    // when recording this information persistently.
+  override def validateAndMaybeEvolveStateSchema(
+      hadoopConf: Configuration,
+      batchId: Long,
+      stateSchemaVersion: Int):
+    Array[String] = {
+    assert(stateSchemaVersion >= 3)
+    val newColumnFamilySchemas = getColFamilySchemas()
+    val schemaFile = new StateSchemaV3File(
+      hadoopConf, stateSchemaFilePath(StateStoreId.DEFAULT_STORE_NAME).toString)
+    // TODO: Read the schema path from the OperatorStateMetadata file
+    // and validate it with the new schema
+
+    // Write the new schema to the schema file
+    val schemaPath = schemaFile.addWithUUID(batchId, newColumnFamilySchemas.values.toList)
+    Array(schemaPath.toString)
+  }
+
+  private def validateSchemas(
+      oldSchemas: List[ColumnFamilySchema],
+      newSchemas: Map[String, ColumnFamilySchema]): Unit = {
+    oldSchemas.foreach { case oldSchema: ColumnFamilySchemaV1 =>
+      newSchemas.get(oldSchema.columnFamilyName).foreach { newSchema =>
+        if (oldSchema != newSchema) {
```

Comment:
Can we reuse some of the checks that the `StateSchemaCompatibilityChecker` does ?